packages/melos/lib/src/common/utils.dart

```diff
@@ -205,6 +209,77 @@ class PromptException extends MelosException {
   String toString() => 'Was unable to prompt for input:\n$message';
 }
 
+class PersistentShell {
+  PersistentShell({
+    required MelosLogger logger,
+    String? workingDirectory,
+  }) {
+    _logger = logger;
+    _workingDirectory = workingDirectory;
+  }
+
+  final _isWindows = currentPlatform.isWindows;
+  late final MelosLogger _logger;
+  late final String? _workingDirectory;
+  late Process _process;
+  /// This list is intended to store the commands that are sent to the shell.
+  /// Currently, it is not being utilized in the code,
+  ///TODO: remove this or actually use it
+  final List<String> _commands = [];
+
+  Future<void> startShell() async {
+    final executable = _isWindows ? 'cmd.exe' : '/bin/sh';
+
+    _process = await Process.start(
+      executable,
+      _isWindows ? ['/C'] : [],
+      workingDirectory: _workingDirectory,
+    );
+
+    _process.stdout.listen(
+      (event) {
+        _logger.logWithoutNewLine(utf8.decode(event, allowMalformed: true));
+      },
+      onDone: () {
+        /// TODO: Identify and log the specific steps that have been completed
+        /// successfully.
+        _logger.success('Shell process completed some steps successfully.');
+      },
+    );
+    _process.stderr.listen(
+      (event) {
+        _logger.error(utf8.decode(event, allowMalformed: true));
+      },
+      onDone: () {
+        /// TODO: Identify and log the specific steps that have completed
+        /// with errors.
+        _logger.error('Shell process completed with errors.');
+      },
+    );
+  }
+
+  void sendCommand(String command) {
+    _commands.add(command);
+
+    final formattedScriptStep = targetStyle(
+      command.addStepPrefixEmoji().withoutTrailing('\n'),
+    );
+
+    final echoCommand = 'echo "$formattedScriptStep"';
+
+    final fullCommand = _isWindows
+        ? '$echoCommand && $command'
+        : 'eval "$echoCommand && $command"';
+
+    _process.stdin.writeln(fullCommand);
```

Comment:
I think after each command has run we have to check the exit code, I'm not sure how we would do that though.
The only easy way I can think of is to store the exit code to a file, but that feels like a very inefficient way...

Reply:
I guess since all of `stdout` is captured we could also echo the error codes and just not print them to the user,
so `echo $?` for *nix and `echo %ERRORLEVEL%` for windows and then just stop the process if a non-zero result is detected.

Reply:
It makes sense, and it also worked out. Thank you for the suggestion. As for stopping the process if a non-zero result is detected, I'm not entirely sure about it, as that's not the way it's working right now. However, if the you guys think is a good idea, we can include it, perhaps as a flag (?). If the user intentionally passes this flag, then the process will stop

Reply:
I think the user should be responsible for letting a command fail and then continue, with a step ending in `|| true` for example. This is how most things like this work as far as I'm aware. :)